core/payments/ondemand/on_demand_ledger_cache.go

```diff
@@ -0,0 +1,183 @@
+	onDemandTableName string
+	// price per symbol in wei, from the PaymentVault
+	pricePerSymbol *big.Int
+	// minimum number of symbols to bill, from the PaymentVault
+	minNumSymbols uint64
+	// protects concurrent access to the ledgers cache during ledger creation
+	//
+	// The lru.Cache object itself is threadsafe, as are the OnDemandLedger values contained in the cache. This lock
+	// is to make sure that only one caller is constructing a new OnDemandLedger at a time. Otherwise, it would be
+	// possible for two separate callers to get a cache miss, create the new object for the same account key, and try
+	// to add them to the cache.
+	//
+	// This lock is intentionally more restrictive than it needs to be, for the sake of simplicity. It could be
+	// converted to an account-based lock instead of a global creation lock, if it ever becomes a bottleneck.
+	ledgerCreationLock sync.Mutex
+	// monitors the PaymentVault for changes, and updates cached ledgers accordingly
+	vaultMonitor *OnDemandVaultMonitor
+}
+
+func NewOnDemandLedgerCache(
+	ctx context.Context,
+	logger logging.Logger,
+	maxLedgers int,
+	paymentVault payments.PaymentVault,
+	updateInterval time.Duration,
+	dynamoClient *dynamodb.Client,
+	onDemandTableName string,
+) (*OnDemandLedgerCache, error) {
+	cache, err := lru.NewWithEvict(
+		maxLedgers,
+		func(accountAddress gethcommon.Address, _ *OnDemandLedger) {
+			logger.Infof("evicted account %s from LRU on-demand ledger cache", accountAddress.Hex())
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("new LRU cache with evict: %w", err)
+	}
+
+	if paymentVault == nil {
+		return nil, errors.New("payment vault must be non-nil")
+	}
+
+	if dynamoClient == nil {
+		return nil, errors.New("dynamo client must be non-nil")
+	}
+
+	pricePerSymbol, err := paymentVault.GetPricePerSymbol(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("get price per symbol: %w", err)
+	}
+
+	minNumSymbols, err := paymentVault.GetMinNumSymbols(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("get min num symbols: %w", err)
+	}
+
+	ledgerCache := &OnDemandLedgerCache{
+		cache:             cache,
+		paymentVault:      paymentVault,
+		dynamoClient:      dynamoClient,
+		onDemandTableName: onDemandTableName,
+		pricePerSymbol:    new(big.Int).SetUint64(pricePerSymbol),
+		minNumSymbols:     minNumSymbols,
+	}
+
+	// Create the vault monitor with callback functions
+	ledgerCache.vaultMonitor, err = NewOnDemandVaultMonitor(
+		ctx,
+		logger,
+		paymentVault,
+		updateInterval,
+		ledgerCache.GetAccountsToUpdate,
+		ledgerCache.UpdateTotalDeposit,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("new on-demand vault monitor: %w", err)
+	}
+
+	return ledgerCache, nil
+}
+
+// Retrieves an existing OnDemandLedger for the given account, or creates a new one if it doesn't exist
+func (c *OnDemandLedgerCache) GetOrCreate(ctx context.Context, accountID gethcommon.Address) (*OnDemandLedger, error) {
```

Comment:
Potential race condition:

- create ledger for client A and hold onto it for a while
- receive a lot of requests, cache is blasted away
- create another ledger for client A
- Two ledgers exist for client A, both ledgers are updated concurrently